// Copyright 2024 Flower Labs GmbH. All Rights Reserved.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
// ==============================================================================

syntax = "proto3";

package flwr.proto;

import "flwr/proto/log.proto";
import "flwr/proto/message.proto";
import "flwr/proto/run.proto";
import "flwr/proto/fab.proto";

service SimulationIo {
  // Pull Simulation inputs
  rpc PullSimulationInputs(PullSimulationInputsRequest)
      returns (PullSimulationInputsResponse) {}

  // Push Simulation outputs
  rpc PushSimulationOutputs(PushSimulationOutputsRequest)
      returns (PushSimulationOutputsResponse) {}

  // Update the status of a given run
  rpc UpdateRunStatus(UpdateRunStatusRequest)
      returns (UpdateRunStatusResponse) {}

  // Push ServerApp logs
  rpc PushLogs(PushLogsRequest) returns (PushLogsResponse) {}

  // Get Federation Options
  rpc GetFederationOptions(GetFederationOptionsRequest)
      returns (GetFederationOptionsResponse) {}
}

// PullSimulationInputs messages
message PullSimulationInputsRequest {}
message PullSimulationInputsResponse {
  Context context = 1;
  Run run = 2;
  Fab fab = 3;
}

// PushSimulationOutputs messages
message PushSimulationOutputsRequest {
  uint64 run_id = 1;
  Context context = 2;
}
message PushSimulationOutputsResponse {}
